src/cp/common.h

```diff
@@ -0,0 +1,156 @@
+// SPDX-FileCopyrightText: 2025 RizinOrg <info@rizin.re>
+// SPDX-FileCopyrightText: 2025 Siddharth Mishra <admin@brightprogrammer.in>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#ifndef CPDEM_COMMON_H
+#define CPDEM_COMMON_H
+
+#include "cp/demangle.h"
+#include "cp/param.h"
+#include "cp/vec.h"
+#include "demangler_util.h"
+
+/**
+ * \b String iterator
+ **/
+typedef struct StrIter {
+    const char* beg; /**< \b Beginning position of string. */
+    const char* end; /**< \b Ending of string (usually points to the null-terminator char). */
+    const char* cur; /**< \b Current read position. */
+} StrIter;
+
+typedef Vec (DemString) ClassNameVec;
+
+typedef struct {
+    StrIter      original;
+    CpDemOptions opts;         // TODO: use options to simplify output while demangling
+
+    ClassNameVec qualifiers;
+    DemString    base_name;    // Used to identify base type in list of function params
+    ParamVec     func_params;  // Names of all function params
+    bool         has_params;   // There are cases where control never reaches `cpdem_func_params`
+
+    DemString suffix;          // anything that is to be put at the very end of demangled output
+    DemString prefix;          // a return type, or another keyword to be put before name
+    DemString custom_operator; // A special case of operator "__op<L><TYPE>"
+
+    bool has_global_name;      /* __vt or _GLOBAL$I/D$, etc... */
+    bool is_ctor;
+    bool is_dtor;
+    ut8  operator_type;        // 0 if not an operator, otherwise a positive value
+} CpDem;
+
+CpDem* cpdem_init (CpDem* dem, const char* mangled, CpDemOptions opts);
+CpDem* cpdem_deinit (CpDem* dem);
```

Comment:
why returning and using as argument the same struct?

Reply:
Returns NULL if fails for any reason.

Reply:
make it a boolean.

Reply:
rename `deinit` as `fini` and make it return `void`